<?xml version="1.0"?>
<robot name="diffbot" xmlns:xacro="http://ros.org/wiki/xacro">

    <link name="map"/>
    <link name="base_link">
        <visual>
            <geometry>
                <box size="0.4 0.8 0.3"/>
            </geometry>
            <material name="blue">
                <color rgba="0 0 .8 1"/>
            </material>
        </visual>
    </link>
    <link name="left_wheel_front">
        <visual>
            <geometry>
                <cylinder length="0.1" radius="0.15"/>
            </geometry>
        </visual>
    </link>
    <link name="left_wheel_rear">
        <visual>
            <geometry>
                <cylinder length="0.1" radius="0.15"/>
            </geometry>
        </visual>
    </link>
    <link name="right_wheel_front">
        <visual>
            <geometry>
                <cylinder length="0.1" radius="0.15"/>
            </geometry>
        </visual>
    </link>
    <link name="right_wheel_rear">
        <visual>
            <geometry>
                <cylinder length="0.1" radius="0.15"/>
            </geometry>
        </visual>
    </link>
    <link name="base_scan">
        <visual>
            <geometry>
                <cylinder length="0.15" radius="0.1"/>
            </geometry>
        </visual>
    </link>
    <link name="odom"/>
    <link name="base_footprint">
        <visual>
            <origin rpy= "0 0 0" xyz = "0 0 0"/>
        </visual>
    </link>

    <!-- Joints -->
    <joint name="map_to_odom" type="fixed">
        <parent link="map"/>
        <child link="odom"/>
        <origin rpy="0 0 1.57" xyz="0 0 0"/>
    </joint>
    <joint name="odom_to_base_link" type="fixed">
        <parent link="odom"/>
        <child link="base_link"/>
        <origin rpy="0 0 0" xyz="0 0 0"/>
    </joint>
    
    <joint name="base_footprint_joint" type="fixed">
        <parent link="base_link"/>
        <child link="base_footprint"/>
        <origin rpy="0 0 0" xyz="0 0 0"/>
    </joint>

    <joint name="lf_wheel_joint" type="continuous">
        <parent link="base_link" />
        <child link="left_wheel_front" />
        <axis xyz="0 1 0" />
        <dynamics damping="0.2"/>
        <limit effort="100" velocity="2.0"/>
        <origin rpy="0 -1.57 0" xyz="-0.4 0.275 -0.1"/>
    </joint>

    <joint name="lr_wheel_joint" type="continuous">
        <parent link="base_link" />
        <child link="left_wheel_rear" />
        <axis xyz="0 1 0" />
        <dynamics damping="0.2"/>
        <limit effort="100" velocity="2.0"/>
        <origin rpy="0 -1.57 0" xyz="-0.4 -0.275 -0.1"/>
    </joint>

    <joint name="rf_wheel_joint" type="continuous">
        <parent link="base_link" />
        <child link="right_wheel_front" />
        <axis xyz="0 1 0" />
        <dynamics damping="0.2"/>
        <limit effort="100" velocity="2.0"/>
        <origin rpy="0 1.57 0" xyz="0.4 0.275 -0.1"/>
    </joint>

    <joint name="rr_wheel_joint" type="continuous">
        <parent link="base_link" />
        <child link="right_wheel_rear" />
        <axis xyz="0 1 10" />
        <dynamics damping="0.2"/>
        <limit effort="100" velocity="2.0"/>
        <origin rpy="0 1.57 0" xyz="0.4 -0.275 -0.1"/>
    </joint>

    <joint name="scan_joint" type="fixed">
        <parent link="base_link" />
        <child link="base_scan" />
        <origin rpy="0 0 -1.57" xyz="0 0.25 0.1"/>
    </joint>

    <!-- ros2_control setup -->
    <ros2_control name="BroverSystem" type="system">
        <hardware>
            <plugin>brover_hardware/BroverSystemHardware</plugin>
        </hardware>

        <joint name="lf_wheel_joint">
            <command_interface name="velocity" />
            <state_interface name="position" />
            <state_interface name="velocity" />
        </joint>

        <joint name="lr_wheel_joint">
            <command_interface name="velocity" />
            <state_interface name="position" />
            <state_interface name="velocity" />
        </joint>

        <joint name="rf_wheel_joint">
            <command_interface name="velocity" />
            <state_interface name="position" />
            <state_interface name="velocity" />
        </joint>

        <joint name="rr_wheel_joint">
            <command_interface name="velocity" />
            <state_interface name="position" />
            <state_interface name="velocity" />
        </joint>

    </ros2_control>

</robot>
